### ISSUE TITLE
Uninstall (silent) / Uninstall all w/ no package.json modification

### ISSUE THREAD

**Issue by typhonrt** [jspm/jspm-cli#1220]
Greets,

I may have missed this in the docs, but there doesn't appear to be an API or CLI way to uninstall individual packages and leave the package.json file unmodified. 

It would also be handy to have an uninstall all w/ option to not modify the package.json file. 

Doing a lot of a library creation w/ SystemJS now and I am manually removing entries from config.js before checking in all the time. I have a bunch of Gulp tasks that execute JSPM via CLI for handy manipulation from WebStorm (having an uninstall / silent CLI option would be handy):
https://github.com/typhonjs/backbone-es6/blob/master/gulpfile.js#L125

I can of course create a Gulp task to manually edit the config.js file I suppose as I have a .gitignore that excludes jspm_packages. It seems handy to have the above functionality in the API / CLI though.

**Comment by guybedford** [comment]
@typhonrt thanks for the suggestion, I'm not sure exactly what sort of use case this is designed to cover - can you perhaps explain why you need to uninstall items before checking in your config?

**Comment by typhonrt** [comment]
It's actually to clear the config.js path and map section really. I suppose it's kind of a clean code obsession on my part. With a .gitignore jspm_packages are not checked in, so why check in the config.js w/ all the installed dependencies of a particular check out / snapshot in time and not a clean map / path section. This is more for libraries too like my Backbone repos. For the ESDoc plugin though the config.js there is only used in automatic testing.  It's probably not a problem for private projects, but for the public libraries / frameworks I'm working on I'd like things as clean as possible. 

In the case of the Backbone libraries there are no other entries in the map / path section. However, in my framework efforts on top of Backbone though the config.js map section in particular has the auto-populated dependencies + a bunch of custom mapped paths. It's nice to be able to remove the code dependencies and just check in the parts that don't change. Especially since when doing any actions with JSPM will rewrite and alphabetize all the entries which intermixes the paths and code dependencies. It makes it a lot harder to remove any dependency entries by hand.

An option though to just uninstall an individual dependency silently with the JSPM API without removing it from package.json would allow me to create a Gulp task to handle things in bulk. A uninstall all silently option would be icing on the cake. It doesn't seem too difficult to add, but I know you're busy with a lot of things. 

This is in line with another background task that I'm going to get around to soon which is an automatic Gulp task to update package.json dependencies and JSPM dependencies with the latest versions with a push of a button. That does bring up another question. It would be nice if there was a JSPM API method that one could query for a given NPM / Github dependency and get back the latest public version that matches a given rule. A unified way to do that from JSPM for NPM / Github dependencies would be great too. Again kind more for clean code reasons than anything else.

**Comment by guybedford** [comment]
It is important though to have all the configuration present as it comprises the version lock of the application. We don't want to encourage removing this. The second feature sounds like https://github.com/jspm/jspm-cli/issues/706, although that is mainly focused on major bumps. It was also considered to have this as a flag in `jspm update` say `jspm update x --bump-versions` or something like that.

**Comment by typhonrt** [comment]
This conversation and suggestion does fall under "workflow enhancements". At least with the libraries I'm creating I consider package.json enough for locking down versions. I assume that your statement above about version lock of an application implies that said application is also checking in jspm_packages to a repository. That does represent a different workflow from what I've been doing thus far with JSPM. An option for uninstall to no modify package.json shouldn't prevent the workflow you are mentioning and supports the workflow I've been using / developing.

With my efforts it's expected that one would run jspm install after checking out the library repos. This potentially will pull down higher versions of dependencies depending on the rules in the package.json. For the master branches of my library repos that is fine as I make sure to bump the base versions and test. Then when doing releases lock down specific versions in the package.json. 

-----

On the second potential feature I'm more thinking of something like `npm outdated` which can return data on all dependencies then the user / developer can perform the update manually or with a script. IE no need to make JSPM guess or make decisions on what to update. It'd be nice though to get formatted data on what is out of date and what potentially needs updating for NPM and Github dependencies.

**Comment by guybedford** [comment]
`jspm install` with no arguments will install the exact version of dependencies from the config file, allowing `jspm_packages` not be checked in.

Formatted output has been suggested before at https://github.com/jspm/jspm-cli/issues/471.

**Comment by typhonrt** [comment]
Thanks for pointing out the use case of the config file having precedence over package.json when using `jspm install` as that makes more sense and I don't recall reading about that in the documentation. Perhaps it could be useful to add to "production workflow" section to the documentation which describes one or more workflows. 

I actually solved my need to clear config.js by separating any mapped paths into a second config file keeping the main config.js only containing dependencies. I have a gulp task now to commit after lint / basic build test and clearing config.js as necessary. I do this for the master branch of the libraries I'm releasing.

There was a small problem with SystemJS Builder in supporting the second config file loading and I have made a pull request here which details a fix which seems safe:
https://github.com/systemjs/builder/pull/377

----
I'm following 471 now as that indeed is discussion pertaining to my second question. I see you sighted npm-check-updates which is the example I'm drawing inspiration on for something similar for JSPM.
